R/mlPrediction.R

```diff
@@ -269,25 +269,38 @@ is.jaspMachineLearning <- function(x) {
   return(names)
 }
 
+.setJaspScaling <- function(x, centers, scales) {
+  if (nrow(x) == 0) {
+    return(x)
+  }
+  for (col in names(centers)) {
+    x[, col] <- (x[, col] - centers[col]) / scales[col]
+  }
+  return(x)
+}
+
 .mlPredictionReadData <- function(dataset, options, model) {
   if (length(options[["predictors"]]) == 0) {
     dataset <- NULL
   } else {
     dataset <- jaspBase::excludeNaListwise(dataset, options[["predictors"]])
-    if (options[["scaleVariables"]] && length(unlist(options[["predictors"]])) > 0) {
-      dataset <- .scaleNumericData(dataset)
-    }
     # Select only the predictors in the model to prevent accidental double column names
     dataset <- dataset[, which(decodeColNames(colnames(dataset)) %in% model[["jaspVars"]][["decoded"]]$predictors), drop = FALSE]
-    # Ensure the column names in the dataset match those in the training data
-    colnames(dataset) <- .matchDecodedNames(colnames(dataset), model)
-    # Retrieve the training set
-    trainingSet <- model[["explainer"]]$data
-    # Check for factor levels in the test set that are not in the training set
-    .checkForNewFactorLevelsInPredictionSet(trainingSet, dataset, "prediction", model)
-    # Ensure factor variables in dataset have same levels as those in the training data
-    factorColumns <- colnames(dataset)[sapply(dataset, is.factor)]
-    dataset[factorColumns] <- lapply(factorColumns, function(i) factor(dataset[[i]], levels = levels(trainingSet[[i]])))
+    if (NCOL(dataset) > 0) {
+      # Ensure the column names in the dataset match those in the training data
+      colnames(dataset) <- .matchDecodedNames(colnames(dataset), model)
+      # Scale the features with the same scaling as the original dataset
+      if (!is.null(model[["jaspScaling"]])) {
+        dataset <- .setJaspScaling(dataset, model$jaspScaling[[1]], model$jaspScaling[[2]])
+      }
+      # Retrieve the training set
+      trainingSet <- model[["explainer"]]$data
+      # Check for factor levels in the test set that are not in the training set
+      .checkForNewFactorLevelsInPredictionSet(trainingSet, dataset, "prediction", model)
+      # Ensure factor variables in dataset have same levels as those in the training data
+      factorColumns <- colnames(dataset)[sapply(dataset, is.factor)]
+      dataset[factorColumns] <- lapply(factorColumns, function(i) factor(dataset[[i]], levels = levels(trainingSet[[i]])))
```

Comment:
I realize this code is unchanged, but the way levels are matched is based on occurrence, no? Isn't that error-prone?

Reply:
Do you mean matching the column names or the levels themselves?

Reply:
I think because we do `.matchDecodedNames(colnames(dataset), model)` this is not an issue, unless you mean the occurrence of the levels?